In [5]:
import sys
if ".." not in sys.path:
    #sys.path.insert(0, "..")
    sys.path.append('../')
import logging
logging.basicConfig(level='ERROR')
from transformers import logging
logging.set_verbosity_error()
import time
import datetime
import os
import wandb
import wget
import pandas as pd
import numpy as np
import random
import torch
from torch import nn
from torch.utils.data import TensorDataset, DataLoader, RandomSampler, SequentialSampler
from transformers import (AutoModel, AutoTokenizer, AutoModelForSequenceClassification, 
    get_linear_schedule_with_warmup, AutoConfig)
from sklearn import metrics
from sklearn.metrics import (accuracy_score, precision_score, recall_score,
f1_score, cohen_kappa_score, roc_auc_score, confusion_matrix, log_loss,
matthews_corrcoef, average_precision_score)

#Custom modules
import utils
from utils import custom_models, early_stopping, worthiness_checker, constants

In [2]:
os.environ["WANDB_SILENT"] = "true"
os.environ["WANDB_NOTEBOOK_NAME"] = 'TransformersForClaimWorthiness.ipynb'

# Constants
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
parent_dir = os.path.dirname(os.path.abspath(os.getcwd()))
seed_list = [7, 42] # seed_list = [7, 42, 127]
fold_count = 3 #5
patience=5
loss_function = nn.BCELoss()

metric_types = np.dtype(
    [
        ("mAP", float),
        ("auc", float),
        ("accuracy", float),
        ("precision", float),
        ("recall", float),
        ("f1", float),
        ("mcc", float),
        ("log_loss", float),
        ("loss", float)
    ])

In [3]:
constants = constants.Constants()
constants.device = device
constants.parent_dir = parent_dir
constants.seed_list = seed_list
constants.fold_count = fold_count
constants.patience = patience
constants.loss_function = loss_function
constants.metric_types = metric_types

### Training Model

In [ ]:
def get_data_from_file(data_version):
    train_df = pd.read_csv(os.path.join(parent_dir, 'Data','train_english_{}.tsv'.format(data_version)), delimiter='\t')
    test_df = pd.read_csv(os.path.join(parent_dir, 'Data','test_english_{}.tsv'.format(data_version)), delimiter='\t')
    return train_df, test_df

In [ ]:
def get_fold_list(df: pd.DataFrame, fold_count, random_state):
    # Group positive and negative samples for stratified sampling
    sampling_df = df.sample(frac=1, replace=False, random_state=random_state)
    sampling_negative_df = sampling_df[sampling_df['check_worthiness']==0]
    sampling_positive_df = sampling_df[sampling_df['check_worthiness']==1]

    #determine fold length for both classes
    fold_size_for_negatives = sampling_negative_df.shape[0]//fold_count
    fold_size_for_positives = sampling_positive_df.shape[0]//fold_count

    fold_list = []
    for i in range(fold_count):
        fold_for_negatives = sampling_positive_df.iloc[fold_size_for_positives*i : fold_size_for_positives * (i+1), :]
        fold_for_positives = sampling_negative_df.iloc[fold_size_for_negatives*i : fold_size_for_negatives * (i+1), :]
        fold_df = pd.concat([fold_for_negatives, fold_for_positives]).sample(frac=1, replace=False, random_state=random_state)

        fold_list.append(fold_df)

    return fold_list

In [ ]:


def create_dataset(df, tokenizer, config):
    max_token_length = config.max_token_length

    sentences = df.tweet_text.values
    labels = df.check_worthiness.values

    # Tokenize all of the sentences and map the tokens to thier word IDs.
    input_ids = []

    for sent in sentences:
        encoded_dict = tokenizer.encode_plus(
                            sent,                      # Sentence to encode.
                            add_special_tokens = True, # Add '[CLS]' and '[SEP]' or equivalent
                            max_length = max_token_length,           # 64? 4-128? Pad & truncate all sentences.
                            truncation=True,
                            padding = 'max_length',
                            return_attention_mask = False,   # Do not Construct attn. masks.
                            return_tensors = 'pt',     # Return pytorch tensors.
                       )

        # Add the encoded sentence to the list.    
        input_ids.append(encoded_dict['input_ids'])

    # Convert the lists into tensors.
    input_ids = torch.cat(input_ids, dim=0)
    labels = torch.tensor(labels).float()
    dataset = TensorDataset(input_ids, labels)
    return dataset

In [ ]:
sweep_config = {
    'name': 'Bert_hyperparameters',
    'method': 'bayes', #grid, random
    'program': 'TransformersForClaimWorthiness.ipynb',
    'early_terminate': {
      'type': 'hyperband',
      'eta': 3,
      's': 2,
      'max_iter': 27   
    },
    'metric': {
      'name': 'avg_val_mAP',
      'goal': 'maximize'   
    },
    'parameters': {
        'data_version': {
          'values': ['raw', 'cleaned_with_mentions', 'cleaned_without_mentions'],
          'distribution': 'categorical'  
        },
        'max_token_length': {
           'min': 4,
           'max': 80,
           'distribution': 'int_uniform'
        },
        'model_name': {
          'values': ['bert-base-uncased'],
          'distribution': 'categorical'  
        },
        'hidden_act': {
          'values': ['relu', 'gelu', 'gelu_new', 'silu'],
          'distribution': 'categorical'  
        },
        'position_embedding_type': {
          'values': ['absolute', 'relative_key', 'relative_key_query'],
          'distribution': 'categorical'  
        },
        'attention_dropout': {
            'min': 0.001,
            'max': 0.2
        },
        'transformer_dropout': {
            'min': 0.001,
            'max': 0.2
        },
        'classifier_dropout': {
            'min': 0.001,
            'max': 0.2
        },
        'layer_norm_eps': {
            'min': 1e-14,
            'max': 1e-10
        },
        'batch_size': {
           'min': 2,
           'max': 80,
           'distribution': 'int_uniform'
        },
        'learning_rate': {
            'min': 0.0000005,
            'max': 0.00025
        },
        'epochs':{
           'min': 1,
           'max': 80,
           'distribution': 'int_uniform'
        }
    }
}


check_points = ['vinai/bertweet-covid19-base-uncased', 'roberta-base', 'bert-base-uncased']

In [ ]:
sweep_defaults = {
    'data_version': 'cleaned_with_mentions',
    'max_token_length': 46,
    'model_name': 'bert-base-uncased',
    'hidden_act': 'gelu',
    'position_embedding_type': 'absolute',
    'layer_norm_eps': 5.4225686692811365e-11,
    'learning_rate': 0.000028734737822604655,
    'transformer_dropout': 0.03873251195245608,
    'attention_dropout': 0.015328152075297112,
    'classifier_dropout': 0.10850207289443518,
    'batch_size': 53,
    'epochs':25
}

In [71]:
# WANDB PARAMETER
def ret_dataloader(batch_size, train_dataset, validation_dataset):

    train_dataloader = DataLoader(
                train_dataset,  # The training samples.
                sampler = RandomSampler(train_dataset), # Select batches randomly
                batch_size = batch_size # Trains with this batch size.
            )

    validation_dataloader = DataLoader(
                validation_dataset, # The validation samples.
                sampler = SequentialSampler(validation_dataset), # Pull out batches sequentially.
                batch_size = batch_size # Evaluate with this batch size.
            )
    return train_dataloader, validation_dataloader

def ret_optim(model, config):
    #print('Learning_rate = ',wandb.config.learning_rate )
    optimizer = torch.optim.AdamW(model.parameters(),
                      lr = config.learning_rate, 
                      eps = 1e-8 
                    )
    return optimizer
    
def ret_scheduler(train_dataloader,optimizer, config):
    epochs = config.epochs
    total_steps = len(train_dataloader) * epochs

    # Create the learning rate scheduler.
    scheduler = get_linear_schedule_with_warmup(optimizer, 
                                                num_warmup_steps = 0, # Default value in run_glue.py
                                                num_training_steps = total_steps)
    return scheduler

def format_time(elapsed):
    '''
    Takes a time in seconds and returns a string hh:mm:ss
    '''
    # Round to the nearest second.
    elapsed_rounded = int(round((elapsed)))
    
    # Format as hh:mm:ss
    return str(datetime.timedelta(seconds=elapsed_rounded))

In [ ]:
def create_dataloaders(fold_list, fold_index, tokenizer, config):
    temp_list = fold_list.copy()

    trial_validation_df = temp_list.pop(fold_index)
    trial_train_df = pd.concat(temp_list)

    train_dataset = create_dataset(trial_train_df, tokenizer, config)
    validation_dataset = create_dataset(trial_validation_df, tokenizer, config)

    return ret_dataloader(config.batch_size, train_dataset, validation_dataset)

In [ ]:

def get_metrics(probability, label_list):
    metrics_dictionary = {}
    # predictions = np.argmax(probability.detach().cpu().numpy(), axis=0)
    predictions =  [int(i > .5) for i in probability]

    accuracy = accuracy_score(label_list, predictions)
    precision = precision_score(label_list, predictions, zero_division=0)
    recall = recall_score(label_list, predictions, zero_division=0)
    f1 = f1_score(label_list, predictions, zero_division=0)
    log_loss = metrics.log_loss(label_list, predictions)
    mcc = matthews_corrcoef(label_list, predictions)
    auc = roc_auc_score(label_list, probability)
    
    mAP = average_precision_score(label_list, probability)

    metric_df = pd.DataFrame(np.empty(0, dtype=metric_types))
    metric_df.loc[0] = [mAP, auc, accuracy, precision, recall, f1, mcc, log_loss, 0]

    return metric_df

In [ ]:
def log_metrics(metric_df, prefix):
    wandb.log({prefix + 'mAP':metric_df.loc[0, 'mAP'],
            prefix + 'auc':metric_df.loc[0, 'auc'],
            prefix + 'accuracy':metric_df.loc[0, 'accuracy'],
            prefix + 'precision':metric_df.loc[0, 'precision'],
            prefix + 'recall':metric_df.loc[0, 'recall'],
            prefix + 'f1':metric_df.loc[0, 'f1'],
            prefix + 'mcc':metric_df.loc[0, 'mcc'],
            prefix + 'log_loss':metric_df.loc[0, 'log_loss'],
            prefix + 'eval_loss':metric_df.loc[0, 'loss']
            })

In [ ]:
def evaluate_one_epoch(model, device, dataloader):
        # Put the model in evaluation mode--the dropout layers behave differently
        # during evaluation.
        model.eval()

        # Tracking variables 
        total_eval_loss = 0
        nb_eval_steps = 0

        probability_list = torch.Tensor(0)
        label_list = np.empty(0)

        # Evaluate data for one epoch
        for batch in dataloader:
            
            b_input_ids = batch[0].to(device)
            b_labels = batch[1].to(device)

            with torch.no_grad():        
                probability = model(b_input_ids).flatten()
                loss = loss_function(probability, b_labels)

            # Accumulate the validation loss, probability and labels.
            total_eval_loss += loss.item()
            probability_list = torch.cat((probability_list, probability.detach().cpu()), axis=0)
            label_list = np.concatenate((label_list, batch[1]), axis=0)

            # Move probability and labels to CPU
            probability = probability.detach().cpu().numpy()
            label_ids = b_labels.to('cpu').numpy()

            # del b_input_ids
            # del b_labels

        # Calculate and log metrics and loss.
        metrics_df = get_metrics(probability_list, label_list)
        metrics_df.loc[0,'loss'] = total_eval_loss / len(dataloader)

        return metrics_df

In [ ]:
def train_one_epoch(model, device, dataloader, loss_function, optimizer, scheduler):
    total_train_loss = 0
    model.train()

    for step, batch in enumerate(dataloader):
        b_input_ids = batch[0].to(device)
        b_labels = batch[1].to(device)

        model.zero_grad()

        probability = model(b_input_ids)
        loss = loss_function(probability.flatten(), b_labels)
        total_train_loss += loss.item()
        loss.backward()

        torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)

        optimizer.step()
        scheduler.step()

        # del b_input_ids
        # del b_labels

    return total_train_loss / len(dataloader)

In [ ]:

def cross_validate():
    # clean gpu memory in any case if previous wandb run was crashed.
    torch.cuda.empty_cache()
    run = wandb.init(config=sweep_defaults)
    run_start_time = time.time()
    # print(wandb.config.items())
    epochs = wandb.config.epochs

    model_name = wandb.config.model_name
    tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=False)

    train_df, test_df = get_data_from_file(wandb.config.data_version)

    run_train_metrics_list = []
    run_val_metrics_list = []

    for seed_index, seed in enumerate(seed_list):

        random.seed(seed)
        np.random.seed(seed)
        torch.manual_seed(seed)
        fold_list = get_fold_list(train_df, 3, random_state=seed)



        for fold_index in range(len(fold_list)): 

            train_dataloader, validation_dataloader = create_dataloaders(fold_list, fold_index, tokenizer, wandb.config)
            model = custom_models.TransformerClassifier(wandb.config).to(device)

            optimizer = ret_optim(model, wandb.config)
            scheduler = ret_scheduler(train_dataloader, optimizer, wandb.config)

            # Creating class that checks early stopping condition
            early_stopping = modules.early_stopping.EarlyStopping(patience=patience)

            epoch_train_metrics_list = []
            epoch_val_metrics_list = []

            for epoch_i in range(0, epochs):

                # ========================================
                #               Training
                # ========================================

                # print("")
                print('======== Seed {:} - Fold {:} - Epoch {:} / {:} ========'.format(seed_index+1, fold_index+1, epoch_i + 1, epochs))
                # print('Training...')
                training_start_time = time.time()

                epoch_train_loss = train_one_epoch(model, device, train_dataloader, loss_function, optimizer, scheduler)

                training_time = format_time(time.time() - training_start_time)
                wandb.log({'train_loss_':epoch_train_loss})

                # print("  Average training loss: {0:.2f}".format(epoch_train_loss))
                #print("  Training epcoh took: {:}".format(training_time))

                # ========================================
                #               Evaluation
                # ========================================

                # print("Running Evaluation...")
                evaluation_start_time = time.time()

                epoch_train_metrics = evaluate_one_epoch(model, device, train_dataloader)
                log_metrics(epoch_train_metrics, 'train_')
                epoch_train_metrics_list.append(epoch_train_metrics)

                epoch_val_metrics = evaluate_one_epoch(model, device, validation_dataloader)
                log_metrics(epoch_val_metrics, 'val_')
                epoch_val_metrics_list.append(epoch_val_metrics)

                val_mAP = epoch_val_metrics['mAP'].loc[0]
                train_mAP = epoch_train_metrics['mAP'].loc[0]

                print("  Training mAP: {:.3f} - Validation mAP: {:.3f}".format(train_mAP,val_mAP ))

                evaluation_time = format_time(time.time() - evaluation_start_time)
                #print("  Evaluation took: {:}".format(evaluation_time))
                if early_stopping.should_stop(val_mAP):
                    # print('terminating because of early stopping!')
                    wandb.log({'early_stopped_at': seed_index*len(fold_list)*epochs+fold_index*epochs+epoch_i + 1})
                    break

            # at the end of each fold, after every epoch finished, 
            # get last epoch`s metrics as final metrics of current fold
            #  # for  df in epoch_val_metrics_list:
            #  #     log_metrics(epoch_val_metrics, 'val_{:}_{:}_'.format(seed_index+1, fold_index+1))

            fold_train_metrics = epoch_train_metrics_list[-1]
            fold_val_metrics = epoch_val_metrics_list[-1]

            run_train_metrics_list.append(fold_train_metrics)
            run_val_metrics_list.append(fold_val_metrics)

            # at the end of every fold, 
            # calculate average metrics as final metrics of current run
            run_train_metrics = pd.concat(run_train_metrics_list)
            run_val_metrics = pd.concat(run_val_metrics_list)

            log_metrics(pd.DataFrame([run_train_metrics.mean()]), 'avg_train_')
            log_metrics(pd.DataFrame([run_val_metrics.mean()]), 'avg_val_')

            # del model
            # torch.cuda.empty_cache()

    # print("")
    # print("Total training took {:} (h:mm:ss)".format(format_time(time.time()-run_start_time)))



In [ ]:
# cross_validate()

In [ ]:
project="Transformers_For_ClaimWorthiness"
entity="cemulu"

In [ ]:
# sweep_id = wandb.sweep(sweep_config, project=project)
sweep_id = 'nbovtee3'

In [ ]:
wandb.agent(sweep_id, project=project,function=cross_validate)

### Training Model with best config and whole training data

In [4]:
api = wandb.Api()
# best_sweep = '2afv0m0i' #bertweet
# best_sweep = 'embywnlj' #roberta
best_sweep = 'nbovtee3' #bert
sweep = api.sweep("cemulu/Transformers_For_ClaimWorthiness/" + best_sweep)

In [5]:
best_run = sweep.best_run()
best_run.summary.get("avg_val_mAP")

wandb: Sorting runs by -summary_metrics.avg_val_mAP


0.7651173954688729

In [6]:
worthiness_checker = worthiness_checker.WorthinessChecker(best_run, constants)

Epoch configuration of the best run:
36
Early stopped at:
                   36    66    123    156    192    240
fold_index         1.0   2.0   3.0    4.0    5.0    6.0
cumulative_epoch  12.0  45.0  90.0  118.0  155.0  195.0
epoch_of_fold     12.0   9.0  18.0   10.0   11.0   15.0

Average epoch used as a reference for early stopping:  8


In [7]:
# optimized_model = worthiness_checker.train_full_model()

model_file_name = 'vinai_bertweet-covid19-base-uncased_0.7651173954688729.pt'
PATH = os.path.join(parent_dir, 'Model', model_file_name)
worthiness_checker.load_model(PATH)

======== Epoch 1 / 8 ========
  Training mAP: 0.845 - Test mAP: 0.660
======== Epoch 2 / 8 ========
  Training mAP: 0.930 - Test mAP: 0.706
======== Epoch 3 / 8 ========
  Training mAP: 0.981 - Test mAP: 0.802
======== Epoch 4 / 8 ========
  Training mAP: 0.995 - Test mAP: 0.670
======== Epoch 5 / 8 ========
  Training mAP: 0.998 - Test mAP: 0.821
======== Epoch 6 / 8 ========
  Training mAP: 1.000 - Test mAP: 0.726
======== Epoch 7 / 8 ========
  Training mAP: 1.000 - Test mAP: 0.750
======== Epoch 8 / 8 ========
  Training mAP: 1.000 - Test mAP: 0.764
*** Training Metrics ***
        mAP       auc  accuracy  precision    recall        f1       mcc  \
0  0.845463  0.895281  0.841849   0.810078  0.720690  0.762774  0.647245   
0  0.929843  0.955542  0.900243   0.866197  0.848276  0.857143  0.780621   
0  0.980538  0.990913  0.951338   0.905844  0.962069  0.933110  0.895926   
0  0.994603  0.998237  0.981752   0.956811  0.993103  0.974619  0.960788   
0  0.997991  0.998982  0.986618   0

In [61]:
model_name = worthiness_checker.config.model_name.replace('/','_')
mAP = best_run.summary.get("avg_val_mAP")

PATH = os.path.join(parent_dir, 'Model','{}_{}.pt'.format(model_name, mAP))
# torch.save(optimized_model.state_dict(), PATH)

### Single Tests

In [35]:
tweet = "UK Health Minister Nadine Dorries has tested positive for COVID-19."

In [10]:
tweet = "i am positive"

In [9]:
tweet = "sheep is black"

In [15]:
tweet = "Recent research suggests that 15 percent of abortions are the result of coercion."

In [23]:
tweet = '''A Democratic bill negotiated between Sens. Joe Manchin and Chuck Schumer would "increase taxes on millions of Americans across every income bracket."'''


In [52]:
tweet = '''Nancy Pelosi and Democrats "want to turn 150 million Americans into felons overnight" with HR 1808.'''

In [32]:
tweet = '''John Fetterman wants to “eliminate life sentences for murderers."'''

In [50]:
tweet = '''"The Sun is out of place, the Moon is out of place and the stars are out of place. The compasses are off" because of a shift in the Earth’s poles.'''

In [109]:
tweet = "China threatens to shoot Nancy Pelosi’s plane down if she visits Taiwan."

In [47]:
tweet = "In Virginia, we actually do protect same-sex marriage."

In [2]:
probability = worthiness_checker.predict(tweet)

NameError: name 'worthiness_checker' is not defined

In [49]:
probability

0.007894571870565414

### Batch Testing

In [67]:
_, test_df = worthiness_checker.get_data_from_file(worthiness_checker.config.data_version)

In [68]:
test_df

tweet_text  check_worthiness
0    POTUS wanted everyone to know he was in close ...                 1
1    Who would you prefer to lead our nation’s resp...                 0
2    It was a really really really really really re...                 0
3    Bald-faced LIE. did self-quarantine until CDC ...                 1
4    LIVE: Daily media briefing on COVID-19 with CO...                 0
..                                                 ...               ...
135  This is how the COVID-19 is spreading and this...                 0
136  Markets are crashing. Tourism is dying. Travel...                 0
137  Pray for Palestine. State of Emergency. - 27 c...                 0
138  ADVISORY TO STUDENTS: Class suspensions were m...                 0
139  The total Iranian COVID-19 case-count is in th...                 1

[140 rows x 2 columns]

In [79]:
test_dataset = worthiness_checker.create_dataset(test_df)
_, test_dataloader = ret_dataloader(worthiness_checker.config.batch_size, _, test_dataset)

In [86]:
def evaluate_one_batch(batch, self):
    b_input_ids = batch[0].to(self.constants.device)
    b_labels = batch[1].to(self.constants.device)
    with torch.no_grad():        
        probability = self.model(b_input_ids).flatten()
        loss = self.constants.loss_function(probability, b_labels)
    probability = probability.detach().cpu()
    return probability, loss

In [87]:
probability_list = torch.Tensor(0)
for batch in test_dataloader:
    probability, _ = evaluate_one_batch( batch, worthiness_checker)
    probability_list = torch.cat((probability_list, probability), axis=0)


In [88]:
probability_list

tensor([0.9889, 0.0067, 0.0076, 0.9908, 0.0075, 0.0071, 0.9879, 0.9914, 0.0065,
        0.9908, 0.9913, 0.9867, 0.9901, 0.9905, 0.9790, 0.0092, 0.0063, 0.9905,
        0.0066, 0.9863, 0.9875, 0.0076, 0.0067, 0.9907, 0.0539, 0.9901, 0.6011,
        0.9912, 0.6066, 0.9866, 0.0067, 0.3569, 0.0078, 0.0072, 0.9907, 0.0078,
        0.3646, 0.0071, 0.0067, 0.0064, 0.0109, 0.9908, 0.9882, 0.9826, 0.0080,
        0.9905, 0.0066, 0.9871, 0.9907, 0.0792, 0.0142, 0.9906, 0.0067, 0.9916,
        0.0087, 0.9911, 0.0064, 0.9911, 0.0067, 0.9878, 0.9899, 0.0075, 0.0063,
        0.0063, 0.0083, 0.9866, 0.0076, 0.9911, 0.0071, 0.0134, 0.0067, 0.9909,
        0.0264, 0.9913, 0.9914, 0.8241, 0.9866, 0.9915, 0.8571, 0.9859, 0.0071,
        0.9863, 0.9897, 0.0068, 0.9432, 0.0069, 0.2583, 0.0066, 0.0070, 0.0081,
        0.0136, 0.6670, 0.0118, 0.9520, 0.9906, 0.0065, 0.0066, 0.9844, 0.0076,
        0.9890, 0.0094, 0.9792, 0.0095, 0.0084, 0.0068, 0.0068, 0.0297, 0.0069,
        0.1405, 0.9886, 0.0071, 0.0065, 

In [91]:
predictions =  [int(i > .5) for i in probability_list]


In [93]:
len(predictions)

140

In [105]:
pd.set_option("display.max_colwidth", 200)

In [107]:
test_df['predictions'] = predictions
test_df['probability'] = probability_list
test_df

tweet_text  \
0              POTUS wanted everyone to know he was in close contact with Gaetz and Collins today, both of whom were supposedly exposed to COVID-19. Did he look worried at the presser? No. It’s a message.   
1                                                                                                                         Who would you prefer to lead our nation’s response to the growing COVID-19 threat?   
2    It was a really really really really really really really really really really really really really really really really really really really really bad idea to elect Donald Trump President of the...   
3                                                                                                       Bald-faced LIE. did self-quarantine until CDC cleared him to return to work. WuFlu COVID-19 COVID-19   
4                                                                                                                                                       LIVE: Daily media briefing on COVID-19 with COVID-19   
..                                                                                                                                                                                                       ...   
135  This is how the COVID-19 is spreading and this is how fast it spreads. We don't have a natural immunity to this virus or a vaccine to protect us. Something as simple as washing your hands can keep...   
136  Markets are crashing. Tourism is dying. Travel is reduced to almost nothing, And World Financial Systems are in chaos. Because of the COVID-19 scare. And you cannot think Of a single reason Why so...   
137  Pray for Palestine. State of Emergency. - 27 cases of COVID-19 - Major churches and mosques closed - No visitors allowed into Palestine - city of Bethlehem under lockdown - Border and Highways clo...   
138  ADVISORY TO STUDENTS: Class suspensions were made for your safety. Please use those days to rest, strengthen your immune system, and read your lessons. Please don't gala muna. :)) WalangPasok COVI...   
139  The total Iranian COVID-19 case-count is in the hundreds of thousands, perhaps millions, according to my estimates (detailed at the link). This raises an important question: if there are two milli...   

     check_worthiness  predictions  probability  
0                   1            1     0.988866  
1                   0            0     0.006661  
2                   0            0     0.007607  
3                   1            1     0.990838  
4                   0            0     0.007499  
..                ...          ...          ...  
135                 0            0     0.009059  
136                 0            0     0.143284  
137                 0            0     0.314902  
138                 0            0     0.006276  
139                 1            0     0.008649  

[140 rows x 4 columns]

False Positives:

* (output = 0.986627) the number of COVID-19 cases in the US surpasses 1,000 with 1,004 people in 37 states and DC testing positive for COVID-19, plus 31 deaths. This is just beginning the acceleration phase of the Cor...

* (output = 0.987125) Italy's Prime Minister Giuseppe Conte has announced that the whole of the country is being put on lockdown in an attempt to contain the COVID-19 outbreak. For the latest on COVID-19, click here:

***

* (output = 0.986261) The empire is striking back. The COVID-19 is now being used as a weapon to destabilize the US economy because that the powers that be feel that’s the only way they can get rid of Trump and regain ...

* (output = 0.857124) As two epidemics - COVID-19 and Brexit - hit us, don’t let them make you forget: Priti Patel scandal and investigations of Johnson’s Arcuri Russia report referendum crimes lies incompetence etc et...

False Negatives

* (output = 0.006674) Democrats and the Media need to stop using the COVID-19 to politicize things and scare people. It's irresponsible. This is not the time to try and gain political points or headlines from scaring p...

* (output = 0.006982) คำขวัญ Thailand 2020 No privacy No security No democracy No hope No future No mask But we had COVID-19 and stupid government Thank you ธนาธร ไม่เอารัฐประหาร

***

* (output = 0.006344) Italian doctor facing COVID-19 tsunami publishes long, moving thread that culminates with, ‘Is panic really worse than neglect and carelessness during an epidemic of this sort?’ Read the whole fri...

* (output = 0.006693) This thread needs to fly. It shows how the legacy media is USING COVID-19 as a political weapon and even how the SAME reporters are contradicting themselves. This. Is. SICK.

Ommitted Link: THREAD: Fri Jan 31, 2020, a few weeks before #Coronavirus has officially spread to other countries (which led to the bad stock market week Feb 24-Feb 28), the Trump Admin announced travel restrictions on China. Here is some of the reporting it generated. Take Politico of 2/4/20.

Tweet: ""private ny colleges: *closed college for the week because of the COVID-19* CUNY: *installed two new hand sanitizer dispensers*""
Label: 0
Prediction: 1 with 98.75% probability

Zero-shot GPT-3 response: "The sentiment of this tweet is that private colleges are not doing enough to prevent the spread of COVID-19, while CUNY is taking steps to protect its students. This is not a check-worthy claim."

In [108]:
test_df[test_df['predictions'] != test_df['check_worthiness']]

tweet_text  \
15   Trust Merkel to be the first world leader to say openly what's what: the main point of social distancing measures (quarantines, closures, lockdowns) is to slow the spread of the COVID-19 so that h...   
19   The empire is striking back. The COVID-19 is now being used as a weapon to destabilize the US economy because that the powers that be feel that’s the only way they can get rid of Trump and regain ...   
20                                                                            private ny colleges: *closed college for the week because of the COVID-19* CUNY: *installed two new hand sanitizer dispensers*   
22                             This thread needs to fly. It shows how the legacy media is USING COVID-19 as a political weapon and even how the SAME reporters are contradicting themselves. This. Is. SICK.   
23   DEVELOPING: confirms directly to that he is self-quarantined after testing negative COVID-19. Meadows has not slowed down, with conference calls preparing for transition WH Chief of Staff. Meadows...   
26   Older adults and people of all ages with severe chronic medical conditions are more likely to develop serious outcomes, including death, if infected with COVID-19. See CDC guidance for people who ...   
29   This was sent to me by a friend from Naples, Italy. The “protezione civile” (civil protection service) has been sending cars (even late at night) with loud speakers urging residents to stay indoor...   
31   Desperate Trump supporters are trying to label COVID-19 the Wuhan virus to distract from the fact that the Trump administration's failures have increased the danger here in the U.S. It's childish,...   
32   The COVID-19 has been around forever. The COVID-19 is a different strain just like there’s different strains of the flu. STOP the panic. Take precaution and follow the guidelines for prevention! m...   
36   On COVID2019, said: “One of the theories is perhaps you could take it on the chin, take it all in one go and allow coronvirus to move through the population without really taking as many draconian...   
38   Democrats and the Media need to stop using the COVID-19 to politicize things and scare people. It's irresponsible. This is not the time to try and gain political points or headlines from scaring p...   
42   US researchers say that it takes an average of 5 days for COVID-19 symptoms (fever, cough, shortness of breath) to appear in COVID-19-infected people. Anyone symptom-free by day 12 is unlikely to ...   
44   Imagine Metro Manila palang may confirmed case. Getting scared kasi di mawari kung paano ang testing process dito sa ibang region Yes guys Philippines is not ready for this kind of situation yet o...   
47        Italy's Prime Minister Giuseppe Conte has announced that the whole of the country is being put on lockdown in an attempt to contain the COVID-19 outbreak. For the latest on COVID-19, click here:   
51             Confirmed cases: 25/2/2020: Italy 320; UK 12 9/3/2020: Italy &gt;7,000; UK 319 Italy don’t have more cases. They’re just two weeks ahead of us. And yet UK govt still doing nothing. COVID-19   
55                   Zee News : Petrol price reduced by Rs 2.69 CNN : Petrol price reduced by Rs 2.69 BBC : Petrol price reduced by Rs 2.69 NDTV : China is sending COVID-19 to world via mails and whatsapp   
56   On a morning when Americans are terrified, the markets are gonna historically crash and we need LEADERSHIP...all you’ve done is hate-tweet BULLSHIT about Sanders, Warren, Biden, Democrats, Schumer...   
59   At CPAC, Mick Mulvaney rattled off his cuckoo COVID-19 hoax theory, claiming liberals and the media are weaponizing COVID-19 to take Trump down. Since then, several CPAC attendees have self-quaran...   
63   Italian doctor facing COVID-19 tsunami publishes long, moving thread that culminates with, ‘Is panic really worse than neglect and carelessness during an epidemic of this sort?’ Read the whole fri...   
75                                

In [1]:
test_df[test_df['predictions'] != test_df['check_worthiness']]

NameError: name 'test_df' is not defined

In [76]:
type(pd.DataFrame())

pandas.core.frame.DataFrame

In [54]:
worthiness_checker.model

TransformerClassifier(
  (transformer): RobertaModel(
    (embeddings): RobertaEmbeddings(
      (word_embeddings): Embedding(64001, 768, padding_idx=1)
      (position_embeddings): Embedding(130, 768, padding_idx=1)
      (token_type_embeddings): Embedding(1, 768)
      (LayerNorm): LayerNorm((768,), eps=3.5894066339034747e-11, elementwise_affine=True)
      (dropout): Dropout(p=0.03312463661668133, inplace=False)
    )
    (encoder): RobertaEncoder(
      (layer): ModuleList(
        (0): RobertaLayer(
          (attention): RobertaAttention(
            (self): RobertaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.0701542558710376, inplace=False)
            )
            (output): RobertaSelfOutput(
              (dense): Linear(in_features=768, out_featu

In [37]:

isCheckWorthy = (probability> .5)
isCheckWorthy

True

In [42]:
print('This expression contains a check-worthy claim with a {:.2%} conficency '.format(probability))

This expression contains a check-worthy claim with a 97.25% conficency 


In [ ]:
optimized_model(input_ids.cuda())

In [ ]:
optimized_model.transformer(input_ids)

In [ ]:
train_df, test_df = get_data_from_file(optimized_config.data_version)
train_dataset = create_dataset(train_df, tokenizer, optimized_config)
test_dataset = create_dataset(test_df, tokenizer, optimized_config)
train_dataloader, test_dataloader = ret_dataloader(optimized_config.batch_size, train_dataset, test_dataset)


In [ ]:
for batch in test_dataloader:
    print(batch)
    break

In [ ]:
model = custom_models.TransformerClassifier(optimized_config)

In [ ]:
model